### ISSUE TITLE
[Compile FE] Build ARM32 target 

### ISSUE THREAD

**Issue by seanshpark** [Samsung/ONE#8379]
Let's enable build compiler frontend for ARM32 target.
- first module to enable is luci
- refer runtime compilation for ARM32
- to enable running inside Tizen

---

Why ARM32 not AArch64?
- Our Tizen target device runs in ARM32.
- From my point of view, ARM32 compilation test is to validate correct execution in ARM32.

**Comment by seanshpark** [comment]
TODO
- [ ] prepare ARM32 rootfs
- [ ] cmake configure
   - [ ] build dependency libraries to ARM32
- [ ] ...

**Comment by seanshpark** [comment]
Issue: flatc native is required to compile fbs file(s)
Issue: protoc, js_embed native is required to compile proto file(s) and generate `well_known_types_js`

**Comment by seanshpark** [comment]
Issue
> compiler/luci/service/include/luci/Service/CircleTypeInference.h:22:10: fatal error: mio/circle/schema_generated.h: No such file or directory
 #include <mio/circle/schema_generated.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by seanshpark** [comment]
For protobuf, `find_package(protobuf EXACT 3.5.2 QUIET)` fails in cross compile.

**Comment by seanshpark** [comment]
Run tests in RPi4
```
sudo apt-get install software-properties-common
sudo apt-get install build-essential cmake scons git
```

**Comment by seanshpark** [comment]
HDF5 uses `TRY_RUN` which doesn't seem to work with cross compile...
Need to find some workarounds.

**Comment by seanshpark** [comment]
> HDF5_CONFIG_DIR-NOTFOUND/hdf5-config.cmake

--> install has problem --> fixed
--> still `find_path` doesn't work for cross build...

**Comment by seanshpark** [comment]
`record-minmax` requires `luci-interpreter`
- need to check kernel codes that use x86 specific codes

**Comment by seanshpark** [comment]
tests are disable as gtest isn't found in cross compile.
- at least I want to run unit tests to run in target device
- later we can enable this in CI with target devices

**Comment by seanshpark** [comment]
How to enable `common-artifacts` ?
- issue: python virtual envionment with TensorFlow 2.6 is needed

Maybe ?
- host drops all reference data
- target executes luci-interpreter and compare with reference data

**Comment by seanshpark** [comment]
Higher version of CMake (Ubuntu 20.04 and Tizen) fails for some `find_package` that was built from source.
`nnas_find_package_folder()` seems to solve this issue.

**Comment by chunseoklee** [comment]
> Higher version of CMake (Ubuntu 20.04 and Tizen) fails for some `find_package` that was built from source. `nnas_find_package_folder()` seems to solve this issue.

This does not solve infinite recursive loop of GTest. I have succeed with the following patch : https://github.com/Samsung/ONE/pull/8481/commits/bd2bb66c3fcb81f09ffd3aa1a518bfb667bc7a79

**Comment by seanshpark** [comment]
> I have succeed with the following patch

OK thanks for the check!

**Comment by seanshpark** [comment]
How to do value test of circle2circle and luci-interpreter with executables from cross build?
- TBD

**Comment by chunseoklee** [comment]
Have you figure out how arm32 cross build succeed w.r.t neon_tensor_utils.cc issue( https://github.com/Samsung/ONE/issues/8480#issuecomment-1047514616 ) ?

**Comment by seanshpark** [comment]
> Have you figure out how arm32 cross build succeed w.r.t neon_tensor_utils.cc

Nope... I'm working on to validate excution result of luci-interpreter.
Currently I'm not sure luci-interpreter for ARM32 build is correct.

**Comment by chunseoklee** [comment]
> Nope... I'm working on to validate excution result of luci-interpreter.
Currently I'm not sure luci-interpreter for ARM32 build is correct.

If luci-interperter is runnable on arm32 device, build semms OK. It is very strange.

**Comment by seanshpark** [comment]
> If luci-interperter is runnable on arm32 device, build semms OK.

I've check the unit test and seems to work OK in RPi4/Ubuntu18.04
I'm working on integration test.

> It is very strange.

Yes, I also don't have any clue...

**Comment by chunseoklee** [comment]
@seanshpark tensorflow lite is built during compilation?

**Comment by chunseoklee** [comment]
> @seanshpark tensorflow lite is built during compilation?

@seanshpark I found that (on my local ubuntu) `-mfpu=neon` flag is not passed in arm32 cross build. Could you please check your compilation command in build/arm32.debug/compilation_commands.json ?

**Comment by seanshpark** [comment]
`luci_value_cross_test` fails only with Sqrt_000 ?

```
      Start 44: luci_value_cross_test
44/45 Test #44: luci_value_cross_test ............***Failed   82.64 sec
FAILED
- Sqrt_000
```

**Comment by seanshpark** [comment]
> I found that (on my local ubuntu) -mfpu=neon flag is not passed in arm32 cross build.

`config_armv7l-linux.cmake` is not include while configuration. I've found that there are missing changes. With this file included, there are link errors like you got for Tizen build.
I'll first finish landing current draft and then work on enabling Neon in luci-interpreter.

**Comment by seanshpark** [comment]
With neon,
```
2: [ RUN      ] L2Pool2DTest.FloatPaddingSameStride
2: compiler/luci-interpreter/src/kernels/L2Pool2D.test.cpp:209: Failure
2: Value of: extractTensorData<float>(output_tensor)
2: Expected: has 8 elements where
2: element #0 is approximately 3.5 (absolute error <= 9.9999997e-06),
2: element #1 is approximately 6 (absolute error <= 9.9999997e-06),
2: element #2 is approximately 6.5 (absolute error <= 9.9999997e-06),
2: element #3 is approximately 5.7008801 (absolute error <= 9.9999997e-06),
2: element #4 is approximately 2.54951 (absolute error <= 9.9999997e-06),
2: element #5 is approximately 7.2111001 (absolute error <= 9.9999997e-06),
2: element #6 is approximately 8.63134 (absolute error <= 9.9999997e-06),
2: element #7 is approximately 7 (absolute error <= 9.9999997e-06)
2:   Actual: { 3.49999, 5.99999, 6.49999, 5.70087, 2.5495, 7.21109, 8.63132, 6.99999 }, whose element #1 doesn't match, which is -1.14441e-05 from 6
2: [  FAILED  ] L2Pool2DTest.FloatPaddingSameStride (0 ms)
```

**Comment by seanshpark** [comment]
> IDK why this diff results in mismatch.

Using neon + ruy gives this difference.

**Comment by seanshpark** [comment]
done except for build test in CI, which added another issue

**Comment by tomdol** [comment]
I'd like to ask a question related to the following PR https://github.com/Samsung/ONE/pull/14535 where I'm attempting to update the protobuf version.

There used to be a patch applied to the protobuf source and it was introduced as a part of this issue's resolution https://github.com/Samsung/ONE/pull/8505

Since the new version of Protobuf does not contain the `js_embed` library any more, is there any other infrastructure change that should be done so that the external (native) `js_embed` is not set nor used?

**Comment by seanshpark** [comment]
> ... should be done so that the external (native) js_embed is not set nor used?

I do not know.

If you like to upgrade protobuf, which ARM32 build maybe affected, please check ARM32 build
and share the results so that I confirm the upgrade is OK.
